### PR TITLE
column pruning

### PR DESCRIPTION
**Pull request by qingxinhome**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #3238

**What this PR does / why we need it:**
In plan2, the columns used by each node in the plan tree are collected from top to bottom in the way of preorder traversal until the leaf node (table_scan), and then the corresponding table field 'ColDef' is marked with isprne. The columns to be trimmed (that is the columns not used in SQL) will be set to true.

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: proto/plan.proto [hunk 1]
```diff
@@ -249,6 +249,7 @@ message ColDef {
 	bool primary        = 5;
 	int32 pkidx 		= 6;
 	string alias		= 7;
+	bool	isPrune		= 8;
```
Comment (ouyuanning) [inline proto/plan.proto:252]: pls rename to :  is_prune
proto will transform to right name;

File: pkg/sql/plan2/utils.go [hunk 2]
```diff
@@ -435,3 +435,52 @@ func replaceColRefWithNull(expr *plan.Expr) *plan.Expr {
...
+	for _, expr := range exprList {
+		err := collectExpr(expr, collect)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func collectExpr(expr *Expr, collect *ColumnCollect) error {
```
Comment (ouyuanning) [inline pkg/sql/plan2/utils.go:450]: maybe error is not necesssary。

File: pkg/sql/plan2/query_builder.go [hunk 3]
```diff
@@ -89,10 +105,18 @@ func (builder *QueryBuilder) remapAllColRefs(nodeId int32) (map[int64][2]int32,
 		}
 
 	case plan.Node_JOIN:
+		// collect the columns required by the current node
+		curNeedCol := newColumnCollect()
+		err := collectDepColumns(node.OnList, curNeedCol)
+		if err != nil {
+			return nil, err
+		}
+		sumNeedCol := mergeColumnCollect(curNeedCol, preNeedCol)
```
Comment (ouyuanning) [inline pkg/sql/plan2/query_builder.go:114]: maybe merge is not necessary。
just pass preNeedCol to collectDepColumns
like :     collectDepColumns(node.OnList, preNeedCol)

**Review by ouyuanning — COMMENTED** [review]
i think you can code it more simple。

File: pkg/sql/plan2/utils.go (hunk 2 above)
Comment (qingxinhome) [reply pkg/sql/plan2/utils.go:450]: yes,  you are right, Let me revise it

File: proto/plan.proto (hunk 1 above)
Comment (qingxinhome) [reply proto/plan.proto:252]: ok, thank you for your advice

File: pkg/sql/plan2/query_builder.go [hunk 4]
```diff
@@ -59,15 +59,31 @@ func (builder *QueryBuilder) remapExpr(expr *Expr, colMap map[int64][2]int32) er
 	return nil
 }
 
-func (builder *QueryBuilder) remapAllColRefs(nodeId int32) (map[int64][2]int32, error) {
+func (builder *QueryBuilder) remapAllColRefs(nodeId int32, preNeedCol *ColumnCollect) (map[int64][2]int32, error) {
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/sql/plan2/query_builder.go:62]: Good idea, thanks. I'll take your advice

File: pkg/sql/plan2/utils.go [hunk 5]
```diff
@@ -435,3 +435,77 @@ func replaceColRefWithNull(expr *plan.Expr) *plan.Expr {
 
 	return expr
 }
+
+// Collect dependent columns in the plan node
+func collectNodeDepColumns(node *Node, curNeedCol *ColumnCollect) {
```
Comment (ouyuanning) [inline pkg/sql/plan2/utils.go:440]: unused code?

File: pkg/vm/pipeline2/types.go [hunk 6]
```diff
@@ -79,8 +83,113 @@ import (
...
+
+// Fill in the blanks to bat
+func resetBat(bat *batch.Batch, pAttrs []string, pAttrTypes []plan.ColDef, proc *process.Process) *batch.Batch {
+	resVectors := make([]*vector.Vector, len(pAttrs))
+	var index int = 0
+	for i, attr := range pAttrs {
+		if attr == "" {
+			scalarVector := proc.AllocScalarVector(types.PlanTypeToType(*(pAttrTypes[i].Typ)))
+			FillScalarValForVector(scalarVector)
+			scalarVector.Length = 1
```
Comment (nnsgmsone) [inline pkg/vm/pipeline2/types.go:121]: The length cannot be 1, it should be equal to len(bat.Zs)

Comment (qingxinhome) [reply pkg/vm/pipeline2/types.go:121]: ok, I will try it later

**Review by fengttt — COMMENTED** [review]
Please update the current status.   If every worked after newest change, I will give this change a greenlight and call the feature done.

However, I want to point out there is no test -- even though every bvt test passes we really don't know if the plan actually did column pruning.

So a task for you, think about how to test optimizer rule.

**Review by yingfeng — COMMENTED** [review]
Why this pr has modifications on irrelevant modules as hakeeper/logservice/metada/metric ? It must have something wrong during the merging process

**Comment by qingxinhome** [comment]
> Why this pr has modifications on irrelevant modules as hakeeper/logservice/metada/metric ? It must have something wrong during the merging process

It appeared after the merger. Let me check it

**Comment by qingxinhome** [comment]
> Please update the current status. If every worked after newest change, I will give this change a greenlight and call the feature done.
> 
> However, I want to point out there is no test -- even though every bvt test passes we really don't know if the plan actually did column pruning.
> 
> So a task for you, think about how to test optimizer rule.

OK, let me think about it, and then add tests for column clipping

**Review by fengttt — APPROVED** [review]

**Pull request merged**